cohttp/fuzz/fuzz_header.ml

```diff
@@ -0,0 +1,563 @@
+
+(* Note : clean_dup does nothing to already concatenated headers. For
+   example, ["a", "v1,v2"] will be not be cleaned. *)
+let clean_dup_test () =
+  Crowbar.(
+    (* FS *)
+    (* Check that there is no more duplicates (except set-cookie). *)
+    add_test
+      ~name:
+        "All headers name in [h] appears strictly once in [clean_dup h] except \
+         for [set-cookie]" [ headers_gen ] (fun h ->
+        let h = H.remove h "set-cookie" in
+        let h = H.(to_list (clean_dup h)) in
+        let compare_key (k, _) (k', _) = compare k k' in
+        check_eq (List.sort_uniq compare_key h) (List.sort compare_key h));
+    (* FS *)
+    (* forall h, k in list_value_headers.
+       String.concat "," (get_multi_concat h k) = get (clean_dup h) k *)
+    add_test ~name:"[clean_dup] concatenates properly list-value headers"
+      [ headers_gen; list_value_header_gen ] (fun h k ->
+        check_eq H.(get_multi_concat h k) H.(get (clean_dup h) k));
+    (* FS *)
+    (* forall h. clean_dup (clean_dup h) = clean_dup h *)
+    add_test ~name:"[clean_dup] is idempotent" [ headers_gen ] (fun h ->
+        check_eq H.(clean_dup (clean_dup h)) H.(clean_dup h));
+    (* FS *)
+    (* forall h. get_multi (clean_dup h) "set-cookie" = get_multi h "set-cookie"*)
+    add_test ~name:"[clean_dup] does nothing to [set-cookie] headers"
+      [ headers_gen ] (fun h ->
+        check_eq
+          H.(get_multi h "set-cookie")
+          H.(get_multi (clean_dup h) "set-cookie"));
+    (* FS *)
+    (* As the generated header values are only composed of letters (it
+       does not generate concatenated values like "gzip,chunked"), the
+       only cases where there are commas in a value is if [clean_dup]
+       concatenated multiple values.
+
+       This test checks that only one value is kept for non-list-value
+       headers and that this value is the last one. *)
+    add_test
+      ~name:"Only list-value headers can have multiple concatenated values "
+      [ headers_gen ] (fun h ->
+        (* As it is an exception, [set-cookie] is removed. *)
+        let h = H.remove h "set-cookie" in
+        let h' = H.(clean_dup h) in
+        let has_multiple_values v =
+          match String.split_on_char ',' v with
+          | [] | [ _ ] -> false
+          | _ -> true
+        in
+        check_eq true
+          H.(
+            fold
+              (fun k v b ->
+                if Array.mem k list_value_headers then b
+                else if has_multiple_values v then false
+                else b && get h k = Some v)
+              h' true)))
+
+let () =
+  init_with_test ();
+  is_empty_test ();
+  mem_test ();
+  add_test ();
+  to_list_of_list_test ();
+  add_opt_test ();
+  add_unless_exists_test ();
+  add_list ();
+  add_multi ();
+  get_test ();
+  get_multi_test ();
+  get_multi_concat_test ();
+  remove_test ();
+  replace_test ();
+  update_test ();
+  update_all_test ();
+  clean_dup_test ();
+  ()
```

Comment:
This module is fantastic. I am going to point here people that want to get an idea on how to use crowbar!

Reply:
Thanks!